EIPS/eip-4762.md

```diff
@@ -0,0 +1,304 @@
+ * If `leaf_key` is not `None` and ``(address, sub_key, leaf_key)`` is not in `accessed_leaves`, charge `WITNESS_CHUNK_COST` gas and add it to `accessed_leaves`
+
+When a **write** event of `(address, sub_key, leaf_key)` occurs, perform the following checks:
+
+ * If (address, sub_key) is not in edited_subtrees, charge `SUBTREE_EDIT_COST` gas and add that tuple to edited_subtrees.
+ * If leaf_key is not None and `(address, sub_key, leaf_key)` is not in `edited_leaves`, charge `CHUNK_EDIT_COST` gas and add it to `edited_leaves`
+     * Additionally, if there was no value stored at `(address, sub_key, leaf_key)` (ie. the state held None at that position), charge `CHUNK_FILL_COST`
+
+Note that tree keys can no longer be emptied: only the values `0...2**256-1` can be written to a tree key, and 0 is distinct from None. Once a tree key is changed from `None` to not-`None`, it can never go back to `None`.
+
+### Replacement for access lists
+
+We replace EIP 2930 access lists with an SSZ structure of the form:
+
+```python
+class AccessList(Container):
+    addresses: List[AccountAccessList, ACCESS_LIST_MAX_ELEMENTS]
+    
+class AccountAccessList(Container):
+    address: Address32
+    subtrees: List[AccessSubtree, ACCESS_LIST_MAX_ELEMENTS]
+    
+class AccessSubtree(Container):
+    subtree_key: uint256
+    elements: BitVector[256]
+```
+
+## Rationale
+
+### Gas reform
+
+Gas costs for reading storage and code are reformed to more closely reflect the gas costs under the new Verkle tree design. `WITNESS_CHUNK_COST` is set to charge 6.25 gas per byte for chunks, and `WITNESS_BRANCH_COST` is set to charge ~13,2 gas per byte for branches on average (assuming 144 byte branch length) and ~2.5 gas per byte in the worst case if an attacker fills the tree with keys deliberately computed to maximize proof length.
+
+The main differences from gas costs in Berlin are:
+
+ * 200 gas charged per 31 byte chunk of code. This has been estimated to increase average gas usage by ~6-12% suggesting 10-20% gas usage increases at a 350 gas per chunk level).
+ * Cost for accessing adjacent storage slots (`key1 // 256 == key2 // 256`) decreases from 2100 to 200 for all slots after the first in the group,
+ * Cost for accessing storage slots 0…63 decreases from 2100 to 200, including the first storage slot. This is likely to significantly improve performance of many existing contracts, which use those storage slots for single persistent variables.
+
+Gains from the latter two properties have not yet been analyzed, but are likely to significantly offset the losses from the first property. It’s likely that once compilers adapt to these rules, efficiency will increase further.
+
+The precise specification of when access events take place, which makes up most of the complexity of the gas repricing, is necessary to clearly specify when data needs to be saved to the period 1 tree.
+
+## Implementation notes
```

Comment:
This is not a valid top level EIP section, can you please fold it into an existing section?